File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLen = 0, left = 0;
        HashSet<Character> char_set = new HashSet<>();

        for (int right = 0; right < s.length(); right++) {
            while (char_set.contains(s.charAt(right))) {
                char_set.remove(s.charAt(left));
                left++;
            }
            char_set.add(s.charAt(right));
            maxLen = Math.max(maxLen, right - left + 1);
        }
        return maxLen;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function (s) {
    const indexMap = new Array(256).fill(-1);
    let left = 0;
    let right = 0;
    let max = 0;

    while (right < s.length) {
        if (indexMap[s.charCodeAt(right)] !== -1) {
            left = Math.max(indexMap[s.charCodeAt(right)] + 1, left);
        }
        indexMap[s.charCodeAt(right)] = right;
        max = Math.max(max, right - left + 1);
        right++;
    }
    return max;
};